rust-mode.el

```diff
@@ -573,6 +584,46 @@ the desired identifiers), but does not match type annotations \"foo::<\"."
             ((not (looking-at (rx (0+ space) "<")))
 	     (throw 'rust-path-font-lock-matcher match))))))))
 
+(defun rust-next-string-interpolation (limit)
+  "Search forward from point for next Rust interpolation marker
+before LIMIT.
+Set point to the end of the occurrence found, and return match beginning
+and end."
+  (while (search-forward "{{" limit t)
+    ;; skip escapes
+    )
+  (catch 'match
+    (save-excursion
+      (while (search-forward "{" limit t)
+        (let ((start (match-beginning 0)))
+          ;; According to fmt_macros::Parser::next, an opening brace
+          ;; must be followed by an optional argument and/or format
+          ;; specifier, then a closing brace. A single closing brace
+          ;; without a corresponding unescaped opening brace is an
+          ;; error. We don't need to do anything special with
+          ;; arguments, specifiers, or errors, so we only search for
+          ;; the single closing brace.
+          (when (search-forward "}" limit t)
+            (throw 'match (list start (point)))))))))
+
+(defun rust-string-interpolation-matcher (limit)
+  "Match next Rust interpolation marker before LIMIT and set
+match data if found. Returns nil if not within a Rust string."
+  (when (rust-in-str)
+    (let ((match (rust-next-string-interpolation limit))) ; too bad `when-let' is 25+
+      (when match
+        (set-match-data match)
```

Comment:
I was surprised to see `set-match-data` here; the docs say that it should be called with a list returned by `match-data` (but also confusingly document what `match-data` returns, making this use seem maybe ok).

But looking more it seems to me that the loop in `rust-next-string-interpolation` could be replaced by a single call to `re-search-forward`, and this could set the match data as well..?